### ISSUE TITLE
Reduce redundant string allocations

### ISSUE THREAD

**Issue by JustinAzoff** [zeek/zeek#1797]
I have been running various pcaps through zeek built with this patch and counting the output.  Have noticed some very interesting things.

```
diff --git a/src/ZeekString.cc b/src/ZeekString.cc
index a939cde26..e4720bdff 100644
--- a/src/ZeekString.cc
+++ b/src/ZeekString.cc
@@ -142,6 +142,10 @@ void String::Set(const u_char* str, int len, bool add_NUL)
 
        if ( add_NUL )
                b[n] = 0;
+    if ( n > 40 )
+        printf("A Allocating a string of len %d\n", n);
+    else
+        printf("A Allocating a string of len %d: %.*s\n", n, n, str);
 
        use_free_to_delete = false;
        }
@@ -157,6 +161,10 @@ void String::Set(const char* str)
                memcpy(b, str, n + 1);
                final_NUL = true;
                use_free_to_delete = false;
+    if ( n > 40 )
+        printf("B Allocating a string of len %d\n", n);
+    else
+        printf("B Allocating a string of len %d: %.*s\n", n, n, str);
                }
        }
 
@@ -169,6 +177,10 @@ void String::Set(const std::string& str)
        memcpy(b, str.c_str(), n + 1);
        final_NUL = true;
        use_free_to_delete = false;
+    if ( n > 40 )
+        printf("C Allocating a string of len %d\n", n);
+    else
+        printf("C Allocating a string of len %d: %.*s\n", n, n, str.c_str());
        }
 
 void String::Set(const String& str)
```

When ran against a pcap of 1 million https downloads, the following are the top string counts:

```
32000345        A Allocating a string of len 6: ShADad
3000045 A Allocating a string of len 3: SSL
2015550 A Allocating a string of len 32: 
2000063 A Allocating a string of len 0: 
2000031 B Allocating a string of len 64
2000030 A Allocating a string of len 2: 
2000021 A Allocating a string of len 64
1008001 A Allocating a string of len 28: 
1000016 A Allocating a string of len 1: 
1000016 A Allocating a string of len 1157
1000015 A Allocating a string of len 8: 
1000015 B Allocating a string of len 3: md5
1000015 B Allocating a string of len 40: c83034b0adb8524863aaa9530edd450c5fdb27db
1000015 A Allocating a string of len 17: d0:37:45:fb:e3:2b
1000015 A Allocating a string of len 4: SHA1
1000015 B Allocating a string of len 6: sha256
1000015 A Allocating a string of len 196
1000015 A Allocating a string of len 17: 02:42:c0:a8:02:15
1000015 A Allocating a string of len 764
1000015 A Allocating a string of len 3: MD5
1000015 A Allocating a string of len 6: SHA256
1000015 B Allocating a string of len 4: sha1
1000015 A Allocating a string of len 4: X509
1000015 A Allocating a string of len 28: application/x-x509-user-cert
1000015 A Allocating a string of len 1: S
1000015 B Allocating a string of len 32: 7bf22b75f299b09c1cdd7d2a9759abe1
1000015 A Allocating a string of len 2: Sh
999219  A Allocating a string of len 9: ShADadfFr
```

Seeing that output I have the following questions:

 * Why is the history string ShADad allocated 32,000,345 times.  I removed an allocation here https://github.com/zeek/zeek/pull/1793  But I think there is still a lot of unnecessary copying of the Conn history field up into c$history.  Can the Conn history field just be the c$history field without a separate c++ string to ZeekString copy?
 * Why are there 2 million allocations of a zero length string. something in the tls analyer?
 * Why are there millions of 32, 2, and 1 byte strings that appear to be empty? tls analyzer?
 * Why are there repeated allocations of strings like md5,MD5,sha1,SHA1,X509 for each connection?
 * Why is application/x-x509-user-cert allocated for each connection?
 * Why are the mac addresses allocated for each connection ( and formatted via fmt_mac )?

**Comment by JustinAzoff** [comment]
I believe the md5/sha1 etc come from things like this:

```
## :zeek:see:`Files::analyzer_name`.
function Files::__analyzer_name%(tag: Files::Tag%) : string
    %{
    const auto& n = zeek::file_mgr->GetComponentName(zeek::IntrusivePtr{zeek::NewRef{}, tag->AsEnumVal()});
    return zeek::make_intrusive<zeek::StringVal>(n);
    %}
```

which is what fills in f$info$analyzers

Could that just be something like

```
## :zeek:see:`Files::analyzer_name`.
function Files::__analyzer_name%(tag: Files::Tag%) : string
    %{
    return tag->AsStringVal();
    %}
```

that returns a pre-allocated string for the tag that never changes.

**Comment by JustinAzoff** [comment]
ah, I think I figured out the history issue. The change in the other MR helped, but I think the proper fix is along the lines of this:

```
--- a/src/Conn.cc
+++ b/src/Conn.cc
@@ -211,6 +211,14 @@ void Connection::HistoryThresholdEvent(EventHandlerPtr e, bool is_orig, uint32_t
        EnqueueEvent(e, nullptr, GetVal(), val_mgr->Bool(is_orig), val_mgr->Count(threshold));
        }
 
+void Connection::AddHistory(char code)
+       {
+       history += code;
+       if ( conn_val )
+               conn_val->Assign(6, history);
+       }
+
+
 const RecordValPtr& Connection::GetVal()
        {
        if ( ! conn_val )
@@ -272,9 +280,6 @@ const RecordValPtr& Connection::GetVal()
        conn_val->AssignTime(3, start_time); // ###
        conn_val->AssignInterval(4, last_time - start_time);
 
-       if ( history.size() )
-               conn_val->Assign(6, history);
-
        conn_val->SetOrigin(this);
 
        return conn_val;
```

**Comment by JustinAzoff** [comment]
This might be closer, 

```
+void Connection::AddHistory(char code)
+       {
+       history += code;
+       GetVal()->Assign(6, history);
+       }
```

Some tests still fail, but I think the current behavior might be wrong.  currently `new_connection` event has `c$history` empty for a tcp connection, with the change it ends up as `S`, which makes sense and a bunch of baselines fail.

**Comment by vpax** [comment]
(Just to double-check: you mean `new_connection` has this for properly-initialized connections, not for all connections, right?)

**Comment by JustinAzoff** [comment]
Currently the contents of `c$history` in new_connection seem to be not well defined.  I tested a few pcaps:

Single syn packet, new_connection will get raised with history being empty.  
Single synack packet between high ports, history is `H` in new_connection.  
Single random data packet, history is empty.

With my patch that updates the connection record only when (and exactly when) AddHistory is called, I get the histories `S`, `H`, and `D`.. which to me makes sense, as that is the current history at the time of the new_connection event and the documentation does say:

```
## Generated for every new connection. This event is raised with the first
## packet of a previously unknown connection. Zeek uses a flow-based definition
## of "connection" here that includes not only TCP sessions but also UDP and
## ICMP flows.
```

**Comment by vpax** [comment]
Yep that makes sense to me too.

**Comment by timwoj** [comment]
@JustinAzoff Is this something you want to get into a PR?

**Comment by JustinAzoff** [comment]
Yeah.. it's a little improvement but does cut down on a bunch of allocations. the tricky part is a TON of the baselines change

**Comment by 0xxon** [comment]
As the fact that c$history is not well-defined by itself is a problem - I am going to add this to the 5.0 milestone.

**Comment by timwoj** [comment]
I just made the suggested `Conn` change above and it didn't seem like a whole lot of tests failed. Most of the ones that failed are because it appears some of the conn IDs changed. I didn't dig into why. I pushed it to a branch on topic/timw/1797-justins-string-allocations  if you want to look at the diffs.

**Comment by timwoj** [comment]
I finally spent a little bit of time digging into why the Connection IDs change. The change to `AddHistory()` to make it call `GetVal()` causes the UIDs to get initialized from different places. For example, there are two connections in `ntlm.log` from the `scripts.base.protocols.dce-rpc.mapi` test. The log changes like this:

```
--- ntlm.log    2022-04-27 22:00:47.929045788 +0000
+++ /home/tim/zeek/testing/btest/.tmp/scripts.base.protocols.dce-rpc.mapi/ntlm.log      2022-04-27 22:00:35.468723734 +0000
@@ -3,9 +3,9 @@
 #empty_field   (empty)
 #unset_field   -
 #path  ntlm
-#open  2022-04-27-22-00-47
+#open  2022-04-27-22-00-35
 #fields        ts      uid     id.orig_h       id.orig_p       id.resp_h       id.resp_p       username       hostname domainname      server_nb_computer_name server_dns_computer_name        server_tree_name        success
 #types time    string  addr    port    addr    port    string  string  string  string  string  string  bool
-1056991899.591337      CFLRIC3zaTU1loLGxh      192.168.0.173   1073    192.168.0.2     1032    ALeonard       ALEONARD-XP      CNAMIS  SATURN  -       -       -
-1056991898.900518      C37jN32gN3y3AZzyf6      192.168.0.173   1068    192.168.0.2     4997    ALeonard       ALEONARD-XP      CNAMIS  SATURN  -       -       -
-#close 2022-04-27-22-00-47
+1056991899.591337      CV5WJ42jPYbNW9JNWf      192.168.0.173   1073    192.168.0.2     1032    ALeonard       ALEONARD-XP      CNAMIS  SATURN  -       -       -
+1056991898.900518      CqlVyW1YwZ15RhTBc4      192.168.0.173   1068    192.168.0.2     4997    ALeonard       ALEONARD-XP      CNAMIS  SATURN  -       -       -
+#close 2022-04-27-22-00-35
```

Looking at the first UID specifically, in master that UID is used from here:

```
1056991899.591337 set uid CFLRIC3zaTU1loLGxh
 0# zeek::UID::Set(unsigned long, unsigned long const*, unsigned long) in zeek
 1# zeek::Connection::GetVal() in zeek
 2# zeek::BifEvent::enqueue_ntlm_negotiate(zeek::analyzer::Analyzer*, zeek::Connection*, zeek::IntrusivePtr<zeek::Val>) in zeek
 3# binpac::NTLM::NTLM_Conn::proc_ntlm_negotiate(binpac::NTLM::NTLM_Negotiate*) in zeek
 4# binpac::NTLM::NTLM_Negotiate::Parse(unsigned char const*, unsigned char const*, binpac::NTLM::ContextNTLM*, int) in zeek
 5# binpac::NTLM::NTLM_SSP_Token::Parse(unsigned char const*, unsigned char const*, binpac::NTLM::ContextNTLM*) in zeek
```

In the branch above that UID is used here:

```
1056991897.498764 set uid CFLRIC3zaTU1loLGxh
 0# zeek::UID::Set(unsigned long, unsigned long const*, unsigned long) in zeek
 1# zeek::Connection::GetVal() in zeek
 2# zeek::Connection::AddHistory(char) in zeek
 3# zeek::Connection::CheckHistory(unsigned int, char) in zeek
 4# zeek::analyzer::tcp::TCP_Endpoint::CheckHistory(unsigned int, char) in zeek
 5# 0x000055F1FE84F7A3 in zeek
 6# zeek::packet_analysis::TCP::TCPSessionAdapter::Process(bool, tcphdr const*, int, std::shared_ptr<zeek::IP_Hdr> const&, unsigned char const*, int) in zeek
 7# zeek::packet_analysis::TCP::TCPAnalyzer::DeliverPacket(zeek::Connection*, double, bool, int, zeek::Packet*) in zeek
```

Basically, I don't think this is a huge deal as long as we're really clear that the IDs might have changed in the NEWS entry for this change.

**Comment by timwoj** [comment]
> ```
> ## :zeek:see:`Files::analyzer_name`.
> function Files::__analyzer_name%(tag: Files::Tag%) : string
>     %{
>     return tag->AsStringVal();
>     %}
> ```

Unfortunately, `Tag` is technically an `EnumVal` here. There's still going to a penalty for lookup in the `ComponentManager` (from `EnumVal` to `Tag`) for every call to `__analyzer_name`, but we can definitely generate the name once there and store it off to be returned on subsequent calls.
